src/Socket/chats.ts

```diff
@@ -1089,24 +1089,40 @@ export const makeChatsSocket = (config: SocketConfig) => {
 			)
 		}
 
-		// State machine: start buffering on pending notifications
-		if (receivedPendingNotifications && syncState === SyncState.Connecting) {
-			syncState = SyncState.AwaitingInitialSync
-			logger.info('Connection is now AwaitingInitialSync, buffering events')
-			ev.buffer()
+		if (!receivedPendingNotifications || syncState !== SyncState.Connecting) {
+			return
+		}
 
-			if (awaitingSyncTimeout) {
-				clearTimeout(awaitingSyncTimeout)
-			}
+		syncState = SyncState.AwaitingInitialSync
+		logger.info('Connection is now AwaitingInitialSync, buffering events')
+		ev.buffer()
 
-			awaitingSyncTimeout = setTimeout(() => {
-				if (syncState === SyncState.AwaitingInitialSync) {
-					logger.warn('Timeout in AwaitingInitialSync, forcing state to Online and flushing buffer')
-					syncState = SyncState.Online
-					ev.flush()
-				}
-			}, 20_000)
+		const willSyncHistory = shouldSyncHistoryMessage(
+			proto.Message.HistorySyncNotification.fromObject({
+				syncType: proto.HistorySync.HistorySyncType.RECENT
+			})
+		)
+
+		if (!willSyncHistory) {
+			logger.info('History sync is disabled by config, not waiting for notification. Transitioning to Online.')
+			syncState = SyncState.Online
+			setTimeout(() => ev.flush(), 0)
```

Comment:
why we need this 0ms timeout?

Reply:
this is to defer the flush to execute after we apply the current pending states